src/treelearner/cuda/cuda_best_split_finder.cu

```diff
@@ -0,0 +1,1804 @@
+      cuda_best_split_info->right_sum_hessians = sum_right_hessian;
+      cuda_best_split_info->right_count = right_count;
+      cuda_best_split_info->left_value = left_output;
+      cuda_best_split_info->left_gain = CUDALeafSplits::GetLeafGainGivenOutput<USE_L1>(sum_left_gradient,
+        sum_left_hessian, lambda_l1, l2, left_output);
+      cuda_best_split_info->right_value = right_output;
+      cuda_best_split_info->right_gain = CUDALeafSplits::GetLeafGainGivenOutput<USE_L1>(sum_right_gradient,
+        sum_right_hessian, lambda_l1, l2, right_output);
+    }
+  } else {
+    __shared__ double shared_value_buffer[NUM_THREADS_PER_BLOCK_BEST_SPLIT_FINDER];
+    __shared__ int16_t shared_index_buffer[NUM_THREADS_PER_BLOCK_BEST_SPLIT_FINDER];
+    __shared__ uint16_t shared_mem_buffer_uint16[32];
+    __shared__ double shared_mem_buffer_double[32];
+    __shared__ int used_bin;
+    l2 += cat_l2;
+    uint16_t is_valid_bin = 0;
+    int best_dir = 0;
+    double best_sum_left_gradient = 0.0f;
+    double best_sum_left_hessian = 0.0f;
+    if (threadIdx_x >= bin_start && threadIdx_x < bin_end) {
+      const int bin_offset = (threadIdx_x << 1);
+      const double hess = feature_hist_ptr[bin_offset + 1];
+      if (__double2int_rn(hess * cnt_factor) >= cat_smooth) {
+        const double grad = feature_hist_ptr[bin_offset];
+        shared_value_buffer[threadIdx_x] = grad / (hess + cat_smooth);
+        is_valid_bin = 1;
+      } else {
+        shared_value_buffer[threadIdx_x] = kMaxScore;
+      }
+    } else {
+      shared_value_buffer[threadIdx_x] = kMaxScore;
+    }
+    shared_index_buffer[threadIdx_x] = threadIdx_x;
+    __syncthreads();
+    const int local_used_bin = ShuffleReduceSum<uint16_t>(is_valid_bin, shared_mem_buffer_uint16, blockDim.x);
+    if (threadIdx_x == 0) {
+      used_bin = local_used_bin;
+    }
+    __syncthreads();
+    BitonicArgSort_1024<double, int16_t, true>(shared_value_buffer, shared_index_buffer, bin_end);
+    __syncthreads();
+    const int max_num_cat = min(max_cat_threshold, (used_bin + 1) / 2);
+
+    if (USE_RAND) {
+      rand_threshold = 0;
+      const int max_threshold = max(min(max_num_cat, used_bin) - 1, 0);
+      if (max_threshold > 0) {
+        rand_threshold = cuda_random->NextInt(0, max_threshold);
+      }
+    }
+
+    // left to right
+    double grad = 0.0f;
+    double hess = 0.0f;
+    if (threadIdx_x < used_bin && threadIdx_x < max_num_cat) {
+      const int bin_offset = (shared_index_buffer[threadIdx_x] << 1);
+      grad = feature_hist_ptr[bin_offset];
+      hess = feature_hist_ptr[bin_offset + 1];
+    }
+    if (threadIdx_x == 0) {
+      hess += kEpsilon;
+    }
+    __syncthreads();
+    double sum_left_gradient = ShufflePrefixSum<double>(grad, shared_mem_buffer_double);
+    __syncthreads();
+    double sum_left_hessian = ShufflePrefixSum<double>(hess, shared_mem_buffer_double);
+    // TODO(shiyu1994): constrain the split with min_data_in_group
```

Comment:
Remove this and #5062 is opened for recording. This should be implemented in a separate PR.